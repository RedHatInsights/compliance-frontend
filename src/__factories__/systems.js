import { Factory } from 'fishery';
import { faker } from '@faker-js/faker';
import { id, hostname } from './helpers';
import { testResultProfiles, buildNonCompliantProfiles } from './profiles';

const systemsFactory = Factory.define(({ params, transientParams }) => {
  const [osMajorVersion, osMinorVersion] = transientParams.osVersions
    ? faker.helpers.arrayElement(transientParams.osVersions).split('.')
    : [];

  return {
    ...id(),
    name: hostname(),
    ...(osMajorVersion
      ? {
          osMajorVersion,
          osMinorVersion,
        }
      : {}),
    ...params,
  };
});

export const buildNonCompliantSystems = (count = 5) =>
  systemsFactory.buildList(count, {
    compliant: false,
    testResultProfiles: [
      ...buildNonCompliantProfiles(),
      ...testResultProfiles.buildList(5),
    ],
  });

export const buildUnsupportedSystems = (count = 5) =>
  systemsFactory.buildList(count, {
    compliant: false,
    testResultProfiles: [...buildNonCompliantProfiles(2, { supported: false })],
  });

export const buildSystemsWithTestResultProfiles = (count = 5, params = {}) =>
  systemsFactory.buildList(count, {
    compliant: params.compliant || false,
    testResultProfiles: [
      ...(!params.compliant ? buildNonCompliantProfiles() : []),
      ...testResultProfiles.buildList(5),
    ],
    ...params,
  });

export default systemsFactory;
