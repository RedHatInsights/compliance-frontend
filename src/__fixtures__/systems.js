export const systems = [
  {
    id: 'd5bc2459-21ce-4d11-bc0b-03ea7513dfa6',
    display_name: 'demo.example.com',
    end_time: '2019-11-21T14:32:19Z',
    failed_rule_count: 28,
    supported: true,
    score: 40.0,
    compliant: false,
    security_guide_version: '0.14.3',
    os_major_version: '7',
    os_minor_version: '8',
    system_id: 'd5bc2459-21ce-4d11-bc0b-03ea7513dfa6',
    type: 'test_result',
    groups: [
      {
        id: 'dc925b53-0bee-4ccf-b95a-4b9611362cf7',
        name: 'Ungrouped Hosts',
        ungrouped: true,
      },
    ],
    tags: [],
  },
  {
    id: 'c1f0aef3-66fd-4abc-aa33-ec6755dd39d9',
    display_name: 'foo.example.com',
    end_time: null,
    failed_rule_count: 0,
    supported: true,
    score: 0.0,
    compliant: false,
    security_guide_version: '0.14.5',
    os_major_version: '7',
    os_minor_version: '8',
    system_id: 'c1f0aef3-66fd-4abc-aa33-ec6755dd39d9',
    type: 'test_result',
    groups: [
      {
        id: 'dc925b53-0bee-4ccf-b95a-4b9611362cf7',
        name: 'Ungrouped Hosts',
        ungrouped: true,
      },
    ],
    tags: [],
  },
];
