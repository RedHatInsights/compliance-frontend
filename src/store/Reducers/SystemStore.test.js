import {
    compliant,
    hasOsInfo,
    lastScanned,
    policyNames,
    countOsMinorVersions,
    systemsToInventoryEntities
} from './SystemStore';
import { systems, entities } from '@/__fixtures__/systems.js';

describe('mapping systems to inventory entities', () => {
    it('should return an empty set if there are no systems', () => {
        expect(systemsToInventoryEntities([], entities, false)).toEqual([]);
    });

    it('should return all systems if showAllSystems is true', () => {
        expect(systemsToInventoryEntities([], entities, true).length).toEqual(entities.length);
    });

    it('should only return systems with a matching inventory entity', () => {
        const oneMatchingEntityInventory = [{ id: 'd5bc2459-21ce-4d11-bc0b-03ea7513dfa6', facts: [] }];
        const inventoryEntities = systemsToInventoryEntities(systems, oneMatchingEntityInventory);
        expect(inventoryEntities.length).toBe(1);
        expect(inventoryEntities).toMatchSnapshot();
    });

    it('should only return all systems if all have a matching in the inventory for showAllSystems=false', () => {
        const inventoryEntities = systemsToInventoryEntities(systems, entities, false);
        const systemIds = systems.map(system => system.node.id);
        const systemNames = systems.map(system => system.node.name);
        const systemComplianceScores = [' 10%', ' N/A'];
        expect(inventoryEntities.length).toBe(2);
        expect(inventoryEntities.map(entity => entity.id).sort()).toEqual(systemIds.sort());
        expect(inventoryEntities.map(entity => entity.display_name).sort()).toEqual(systemNames.sort());
        expect(inventoryEntities.map(entity => entity.facts.compliance.compliance_score_text).sort()).
        toEqual(systemComplianceScores.sort());
    });

    it('should always return all systems for showAllSystems=true', () => {
        const inventoryEntities = systemsToInventoryEntities(systems, entities, true);
        const systemComplianceScores = [' 10%', ' N/A', ' N/A', ' N/A'];
        expect(inventoryEntities.length).toBe(entities.length);
        expect(inventoryEntities.map(entity => entity.id).sort()).toEqual(entities.map(e => e.id).sort());
        expect(inventoryEntities.map(entity => entity.display_name).sort()).toEqual(entities.map(e => e.display_name).sort());
        expect(inventoryEntities.map(entity => entity.facts.compliance.compliance_score_text).sort()).
        toEqual(systemComplianceScores.sort());
    });
});

describe('.lastScanned', () => {
    it('should find the latest scan date', () => {
        const system = {
            testResultProfiles: [
                { lastScanned: '2019-10-25T15:59:49Z' },
                { lastScanned: '2019-10-23T15:59:49Z' },
                { lastScanned: '2018-12-23T17:59:49Z' }
            ]
        };
        expect(lastScanned(system)).toEqual(new Date('2019-10-25T15:59:49Z'));
    });

    it('should print the latest scan date even if one profile was never scanned', () => {
        const system = {
            testResultProfiles: [
                { lastScanned: '2019-10-25T15:59:49Z' },
                { lastScanned: 'Never' }
            ]
        };
        expect(lastScanned(system)).toEqual(new Date('2019-10-25T15:59:49Z'));
    });

    it('should print Never if the scan date cannot be ascertained', () => {
        expect(lastScanned({ testResultProfiles: [] })).toEqual('Never');
        expect(lastScanned({ testResultProfiles: [{ lastScanned: 'Never' }] })).toEqual('Never');
    });
});

describe('.compliant', () => {
    it('should set false if there is one non-compliant profile', () => {
        const system = {
            testResultProfiles: [
                { compliant: true },
                { compliant: false }
            ]
        };
        expect(compliant(system)).toEqual(false);
    });

    it('should set true if all profiles are compliant', () => {
        const system = {
            testResultProfiles: [
                { compliant: true },
                { compliant: true }
            ]
        };
        expect(compliant(system)).toEqual(true);
    });
});

describe('.policyNames', () => {
    it('should return all system policies', () => {
        const system = {
            testResultProfiles: [
                { name: 'HIPAA Profile', policy: {} },
                { name: 'PCI-DSS Profile', policy: {} },
                { name: 'CIS', policy: null }
            ],
            policies: [
                { name: 'HIPAA' },
                { name: 'PCI Policy' }
            ]
        };
        expect(policyNames(system)).toEqual('HIPAA, PCI Policy, (External) CIS');
    });
});

describe('hasOsInfo', () => {
    it('return true for 7.3', () => {
        const result = hasOsInfo({
            osMajorVersion: 7,
            osMinorVersion: 3
        });
        expect(result).toEqual(true);
    });

    it('return true for 8.0', () => {
        const result = hasOsInfo({
            osMajorVersion: 8,
            osMinorVersion: 0
        });
        expect(result).toEqual(true);
    });

    it('return false for both versions being 0', () => {
        const result = hasOsInfo({
            osMajorVersion: 0,
            osMinorVersion: 0
        });
        expect(result).toEqual(false);
    });

    it('return false for version being null', () => {
        const result = hasOsInfo({
            osMajorVersion: null,
            osMinorVersion: null
        });
        expect(result).toEqual(false);
    });
});

describe('countOsMinorVersions', () => {
    it('gathers unique OS minor versions sorted', () => {
        const systems = [
            { osMinorVersion: 10 },
            { osMinorVersion: 1 },
            { osMinorVersion: 9 },
            { osMinorVersion: 9 },
            { osMinorVersion: 0 },
            {}
        ];
        expect(countOsMinorVersions(systems)).toEqual([
            { osMinorVersion: 0, count: 1 },
            { osMinorVersion: 1, count: 1 },
            { osMinorVersion: 9, count: 2 },
            { osMinorVersion: 10, count: 1 }
        ]);
    });
});
