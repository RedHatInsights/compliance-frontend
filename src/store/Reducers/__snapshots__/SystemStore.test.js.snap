// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping systems to inventory entities should only return systems with a matching inventory entity 1`] = `
Array [
  Object {
    "account": undefined,
    "bios_uuid": undefined,
    "created": undefined,
    "display_name": undefined,
    "facts": Object {
      "compliance": Object {
        "compliance_score": <React.Fragment>
          <CompliantIcon
            id="d5bc2459-21ce-4d11-bc0b-03ea7513dfa6"
            lastScanned={2019-11-21T14:32:19.000Z}
            name="demo.lobatolan.home"
            profileNames="Standard System Security Profile for Red Hat Enterprise Linux 7, DISA STIG for Red Hat Enterprise Linux 7, United States Government Configuration Baseline, C2S for Red Hat Enterprise Linux 7"
            profiles={
              Array [
                Object {
                  "compliant": false,
                  "lastScanned": "2019-11-21T14:32:19Z",
                  "name": "Standard System Security Profile for Red Hat Enterprise Linux 7",
                  "rulesFailed": 34,
                  "rulesPassed": 17,
                },
                Object {
                  "compliant": false,
                  "lastScanned": "2019-11-21T14:32:19Z",
                  "name": "DISA STIG for Red Hat Enterprise Linux 7",
                  "rulesFailed": 152,
                  "rulesPassed": 65,
                },
                Object {
                  "compliant": false,
                  "lastScanned": "2019-11-21T14:32:19Z",
                  "name": "United States Government Configuration Baseline",
                  "rulesFailed": 190,
                  "rulesPassed": 171,
                },
                Object {
                  "compliant": false,
                  "lastScanned": "2019-11-21T14:32:19Z",
                  "name": "C2S for Red Hat Enterprise Linux 7",
                  "rulesFailed": 129,
                  "rulesPassed": 83,
                },
              ]
            }
            rulesFailed={505}
            rulesPassed={336}
          />
           40%
        </React.Fragment>,
        "compliance_score_text": " 40%",
        "last_scanned": Object {
          "title": <FormattedRelative
            updateInterval={10000}
            value={1574346739000}
          />,
        },
        "last_scanned_text": 2019-11-21T14:32:19.000Z,
        "profiles": "Standard System Security Profile for Red Hat Enterprise Linux 7, DISA STIG for Red Hat Enterprise Linux 7, United States Government Configuration Baseline, C2S for Red Hat Enterprise Linux 7",
        "rules_failed": Object {
          "title": <ForwardRef
            to={
              Object {
                "pathname": "/systems/d5bc2459-21ce-4d11-bc0b-03ea7513dfa6",
                "query": Object {
                  "hidePassed": true,
                },
              }
            }
          >
            505
          </ForwardRef>,
        },
        "rules_failed_text": 505,
        "rules_passed": 336,
      },
      "inventory": Object {
        "hostname": undefined,
        "machine_id": undefined,
        "release": undefined,
      },
    },
    "fqdn": undefined,
    "id": "d5bc2459-21ce-4d11-bc0b-03ea7513dfa6",
    "insights_id": undefined,
    "ip_addresses": undefined,
    "mac_addresses": undefined,
    "rhel_machine_id": undefined,
    "satellite_id": undefined,
    "subscription_manager_id": undefined,
    "tags": undefined,
    "updated": undefined,
  },
]
`;
